core/src/main/java/org/apache/iceberg/CatalogUtil.java

```diff
@@ -275,6 +276,27 @@ public static FileIO loadFileIO(
     return fileIO;
   }
 
+  public static MetricsContext loadFileIOMetricsContext(
+      String impl,
+      Map<String, String> properties,
+      Object hadoopConf) {
+    LOG.info("Loading custom MetricsContext implementation: {}", impl);
+    MetricsContext metricsContext;
+    try {
+      DynConstructors.Ctor<MetricsContext> ctor =
+              DynConstructors.builder(MetricsContext.class).impl(impl).buildChecked();
+      metricsContext = ctor.newInstance();
+    } catch (NoSuchMethodException | ClassCastException e) {
+      LOG.warn("Unable to load metrics class: '{}', falling back to null metrics", impl, e);
+      return null;
```

Comment:
We should return the null metrics impl not `null` here because engines without hadoop (e.g. Flink) will then NPE in the S3FileIO.

Reply:
Specifically `MetricsContext.nullMetrics()`

Reply:
yes good point I missed that, this should not return null

Reply:
Thank you for that. I have fixed that